pkg/fanal/analyzer/language/nodejs/yarn/yarn.go

```diff
@@ -177,3 +210,148 @@ func (a yarnAnalyzer) parsePackageJsonDependencies(fsys fs.FS, path string) (map
 	// Merge dependencies and optionalDependencies
 	return lo.Assign(pkg.Dependencies, pkg.OptionalDependencies), nil
 }
+
+type traverseFunc func(pkg packagejson.Package) error
+
+func (a yarnAnalyzer) traversePkgs(fsys fs.FS, lockPath string, fn traverseFunc) error {
+	dir := filepath.Dir(lockPath)
+
+	nodeModulesPath := path.Join(dir, "node_modules")
+
+	if _, err := fs.Stat(fsys, nodeModulesPath); errors.Is(err, fs.ErrNotExist) {
+		// try to find for yarn v2+
+		return a.traverseYarnModernPkgs(fsys, dir, fn)
+	} else if err != nil {
+		return xerrors.Errorf("unable to parse %q: %w", nodeModulesPath, err)
+	}
+
+	return a.traverseYarnClassicPkgs(fsys, nodeModulesPath, fn)
```

Comment:
I think about 1 case when `node_modules` and `.yarn` folders exist.
E.g. User updated project to yarn v2.
Does `node_modules` folder still exist or will `yarn` remove this folder?

@nikpivkin wdyt? Perhaps we need to check `.yarn` folder first (as latest implementation) or(and) add debug message that 2 folder were found.

Reply:
@DmitriyLewen then the user should delete `node_modules`)
In fact, if the yarn [nodeLinker](https://yarnpkg.com/configuration/yarnrc#nodeLinker) is set as `node_modules`, then the packages will be installed in` node_modules`, but the folder `.yarn` can also exist, because it stores not only packages.

Reply:
Thanks for information.

I also played with `yarn`:
After updating `yarn` and adding/updating package -> `yarn` moves packages from `node_modules` to `.yarn` folder.
Some example:
```
➜  docker run --rm -it node@sha256:2d5e8a8a51bc341fd5f2eed6d91455c3a3d147e91a14298fc564b5dc519c1666 sh

/ # mkdir test && cd "$_"

/test # yarn set version 1.22.19
Resolving 1.22.19 to a url...
Downloading https://github.com/yarnpkg/yarn/releases/download/v1.22.19/yarn-1.22.19.js...
Saving it into /test/.yarn/releases/yarn-1.22.19.cjs...
Updating /test/.yarnrc...
Done!

/test # yarn add is-callable@1.2.7
...

/test # ls -hl node_modules/  
drwxr-xr-x    4 root     root        4.0K Jun 29 04:40 is-callable

/test # ls -hl ./.yarn
drwxr-xr-x    2 root     root        4.0K Jun 29 04:40 releases

/test # yarn set version 3.4.1
...

/test # yarn add is-odd@3.0.1
...

/test # ls -hl node_modules/
ls: node_modules/: No such file or directory

/test # ls -hl .yarn
drwxr-xr-x    2 root     root        4.0K Jun 29 04:41 cache
-rw-r--r--    1 root     root        1.5K Jun 29 04:41 install-state.gz
drwxr-xr-x    2 root     root        4.0K Jun 29 04:41 releases

/test # ls -hl .yarn/cache
total 28K    
-rw-r--r--    1 root     root       12.3K Jun 29 04:41 is-callable-npm-1.2.7-808a303e61-61fd57d03b.zip
-rw-r--r--    1 root     root        4.2K Jun 29 04:41 is-number-npm-6.0.0-30881e83e6-f73bfced02.zip
-rw-r--r--    1 root     root        3.6K Jun 29 04:41 is-odd-npm-3.0.1-93c3c3f41b-4e2b20764d.zip

```